File: Supervised Learning/Multi-class Classification/One-Vs-All Logistic Regression/Octave/oneVsAll.m
function [all_theta] = oneVsAll(X, y, num_labels, lambda)
%ONEVSALL trains multiple logistic regression classifiers and returns all
%the classifiers in a matrix all_theta, where the i-th row of all_theta 
%corresponds to the classifier for label i
%   [all_theta] = ONEVSALL(X, y, num_labels, lambda) trains num_labels
%   logistic regression classifiers and returns each of these classifiers
%   in a matrix all_theta, where the i-th row of all_theta corresponds 
%   to the classifier for label i


% We will use Advanced Optimization custom function fmincg instead of fminunc.
% fmincg works similar to fminunc, but is more efficient for dealing with a large number of parameters.

% In Octave/MATLAB, evaluating the expression a == b for a vector a (of size m×1) and scalar b will return 
% a vector of the same size as a with ones at positions where the elements of a are equal to b and zeroes 
% where they are different. 
% For example: 
%   a = 1:10;
%   b = 3;
%   a == b    
% last statement will return '0  0  1  0  0  0  0  0  0  0'

% We will use this feature to select one vs all for a selected value. 
% For example y==2 will return value of y with 1 set for position of alls 2's and rest will be set to 0.   


%Initialization
m = size(X, 1);
n = size(X, 2);

all_theta = zeros(num_labels, n + 1);

% Add ones to the X data matrix for x0=1
X = [ones(m, 1) X];

initial_theta = zeros(n + 1, 1);
options = optimset('GradObj', 'on', 'MaxIter', 50);

for c = 1:num_labels
	[theta] = fmincg (@(t)(costFunctionReg(t, X, (y == c), lambda)), initial_theta, options);
	%[theta, J, exit_flag] = fminunc(@(t)(costFunctionReg(t, X, (y == c), lambda)), initial_theta, options);
    all_theta(c,:) = theta';
endfor

end

File: Supervised Learning/Multi-class Classification/One-Vs-All Logistic Regression/Python/oneVsAll.py
import numpy as np
import loadData as data
import optimizeTheta as optTheta

def oneVsAll(num_labels, lmda):

	# ONEVSALL trains multiple logistic regression classifiers and returns all
	# the classifiers in a matrix all_theta, where the i-th row of all_theta
	# corresponds to the classifier for label i
	#    [all_theta] = ONEVSALL(X, y, num_labels, lambda) trains num_labels
	#    logistic regression classifiers and returns each of these classifiers
	#    in a matrix all_theta, where the i-th row of all_theta corresponds
	#    to the classifier for label i


	X = np.insert(data.X, 0, 1, axis=1)
	y = data.y

	# Set initial &theta; to zero (3x1 vector)
	initial_theta = np.zeros((X.shape[1], 1))
	all_theta = np.zeros((num_labels, X.shape[1]))

	iterations = 50


	for c in np.arange(1, num_labels + 1):
		thetaOptimized, costOptimized = optTheta.calculateOptimizeTheta(initial_theta, X, (y == c)*1, iterations, lmda=lmda)
		all_theta[c - 1,:] = np.transpose(thetaOptimized)

	return all_theta


def main():

    lmda = 0.1
    num_labels = 10
    theta = oneVsAll(num_labels,lmda)
    print(theta[:4,:1])

    # Expected first 4 values: -3.00407911, -2.95000729, -4.88676598, -2.2890568



# If this script is executed, then main() will be executed
if __name__ == '__main__':
    main()



